### PR TITLE
tough: add method for caching a repo with its metadata

### LINKED ISSUE #80: Add method for saving metadata relevant to a set of targets
I'd like to fetch a target from a TUF repo, store it locally, and also store with it the metadata necessary to form the trust chain from my trusted root metadata file to that target. This would allow me to continue using the target in a trusted way when a machine has not yet configured networking, for instance.

- [ ] Store the root.json trust chain
- [ ] Add a method that returns the set of metadata required to verify a set of targets

What my brain thinks might be a good interface for this right now:

```rust
fn metadata_chain(
    &self,
    targets: impl Iterator<Item = impl AsRef<str>>
) -> impl Iterator<Item = (String, Signed<Box<dyn Role>>)>
```

This is incredibly unwieldy and needs some work :)

### PR DESCRIPTION
**Pull request by webern**
*Issue #, if available:*

Closes #80
Supports  https://github.com/bottlerocket-os/bottlerocket/issues/91
Supports https://github.com/bottlerocket-os/bottlerocket/issues/905

*Description of changes:*

In support of Bottlerocket's datastore migrations, we need to locally cache TUF targets along with the metadata that they were signed with.
This PR adds a function, named `cache` to the `Repository` object, which takes a list of targets, and stores them along with the repo metadata into directories for future use.

*Testing Performed*

Ran existing unit/integ tests and added integ tests for the new cache function.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: tough/src/error.rs
Comment (zmrow) [inline tough/src/error.rs]: Nit: Add quotations around the `Url`/`Pathbuf` in these errors:

`Error reading data from '{}'...`

File: tough/src/lib.rs
Comment (zmrow) [inline tough/src/lib.rs]: Not disagreeing, just asking: Is there any reason to separate these directory arguments?  Is there a reason to put metadata and targets at different locations?

File: tough/src/lib.rs
Comment (zmrow) [inline tough/src/lib.rs]: Could we add a small comment here? At first glance I wondered why you were calling the same function 3 times until I saw it was to cache all the metadata.

File: tough/tests/repo_cache.rs
Comment (zmrow) [inline tough/tests/repo_cache.rs]: s/desintaion/destination/g

**Review by zmrow — CHANGES_REQUESTED** [review]
Overall, it would be super helpful to add some more comments to all the functions that were added. I was getting wrapped around the axle keeping some of them straight.

Sidebar to that thought: I wonder if this might be better implemented separately as a different data structure? Perhaps a `Cache` struct with some methods hanging off of it? Something to think about but I won't block on it. @iliana might have more to add here.

File: tough/src/lib.rs
Comment (webern) [reply tough/src/lib.rs]: It’s not required to do-so, unless you happen to be a) not using consistent snapshots, and b) have a target file with the same name as a metadata file. If a and b are true then you could not save targets and metadata in the same dir.

**Comment by webern** [comment]
> more comments to all the functions that were added. I was getting wrapped around the axle keeping some of them straight.

Agreed, I will liberally add comments based on your feedback. I tend to prefer small helper functions that reduce the overall length and visual complexity of a routine, but not everybody agrees on where to draw the line on such (lingering effects of reading Clean Code a long time ago).

> Perhaps a `Cache` struct with some methods hanging off of it? Something to think about but I won't block on it. @iliana might have more to add here.

I agree that `lib.rs` is growing too big. And, more specifically, that Repository is getting big. Although, the entire library culminates in a Repository object so it might be hard to avoid the god class anti-pattern there.

I would vote against a struct for Cache simply because there is no data that it needs to hold (other than a Repository). I could create a cache.rs file, and extract any/all functions that are only used for the cache function, and which do not require private access into the Repository.

**Comment by zmrow** [comment]
> Agreed, I will liberally add comments based on your feedback. I tend to prefer small helper functions that reduce the overall length and visual complexity of a routine, but not everybody agrees on where to draw the line on such (lingering effects of reading Clean Code a long time ago)

😄 I have no problems with the style, I just find comments are helpful.

File: tough/src/lib.rs
Comment (webern) [inline tough/src/lib.rs]: I discovered this bug after opening the PR. I will fix this. It should be max root size for root.json files and signed size (i think) for other metadata file. So probably max_size needs to be added to the function signature.

**Comment by webern** [comment]
> I have no problems with the style

At least I deleted this function which was ~~kind of~~ ridiculous...

```rust
    fn find_target(&self, name: &str) -> Option<&Target> {
        let targets = &self.targets.signed.targets;
        targets.get(name)
    }
```

**Comment by webern** [comment]
Converted to draft. Properly correcting for use of the correct max_size is harder than I thought.

**Comment by webern** [comment]
Re-opened, now using the correct file size limits. I tried to find a better way of providing the `specifier` string literals, but all solutions were more obfuscating than just passing a string literal. I documented the `specifier` argument because it took me a while to understand exactly what it meant.

File: tough/src/error.rs
Comment (webern) [inline tough/src/error.rs]:
```suggestion
    #[snafu(display("The target '{}' was not found", target_name))]
```

File: tough/src/lib.rs
Comment (iliana) [inline tough/src/lib.rs]:
```suggestion
        targets_subset: Option<&[S]>,
```

... `where S: AsRef<str>`, to make the set of types that can be used here much wider (you can pass something like `Vec<String>` that implicitly `Deref`s to a slice, or a `&[&str]`, or anything in between).

File: tough/src/lib.rs
Comment (iliana) [inline tough/src/lib.rs]: This isn't correct -- timestamp.json doesn't have the version prepended even when consistent_snapshot is true.

**Comment by iliana** [comment]
My other larger feedback is to split out the cache-specific methods. You can have two modules implement methods on the same struct, just make sure the struct is in scope and open a new `impl Repository` block. The current `impl Repository` block is getting *very* big.

Assume you have my approval if the two callouts in my review are fixed. :)

File: tough/src/lib.rs
Comment (webern) [reply tough/src/lib.rs]: Nice, tip.

**Comment by webern** [comment]
Revised https://github.com/awslabs/tough/compare/9d97ad7b4024c007cd7815b6e8930ab7d30617dd..5faa548fba3513c207d519306c3ea7dd268fe17b

**Review by zmrow — APPROVED** [review]
🕹️

**Pull request merged**
